### PR TITLE
test(mcp-oauth): cover token storage, callback handlers, browser detection, and paste reader for 99% coverage

### LINKED ISSUE #36567: Improve test coverage: tools/mcp_oauth.py
## Coverage gap

Module `tools/mcp_oauth.py` is below the 70% coverage threshold.

- Current module coverage: 31.23%
- Statements: 317
- Missing lines/statements: `[70, 71, 75, 76, 121, 122, 134, 135, 142, 143, 149, 150, 152, 153, 154, 155, 156, 157, 158, 160, 161, 162, 169, 170, 171, 172, 173, 200, 201, 202, 203, 204, 208, 209, 261, 262, 263, 264, 265, 266, 267, 268, 269, 270, 271, 272, 273, 274, 275, 276, 277, 278, 279, 294, 297, 299, 307, 308, 309, 310, 311, 314, 315, 326, 327, 333, 334, 335, 336, 337, 343, 344, 364, 366, 367, 368, 369, 370, 371, 373]`
- Project total coverage: 43.22%

## Reproduce

```bash
cd /Users/agentswarm/.hermes/hermes-agent
/Users/agentswarm/.hermes/hermes-agent/venv/bin/python -m pytest tests --cov=. --cov-report=term-missing --cov-report=json:/tmp/coverage-hermes-agent-1780293804.json
```

## Acceptance criteria

- Bring `tools/mcp_oauth.py` to >= 70% statement coverage.
- Keep the coverage/test command above passing.

### PR DESCRIPTION
**Pull request by Christopher-Schulze**
## What does this PR do?

Extends `tests/tools/test_mcp_oauth.py` across token persistence, snapshot and restore recovery, browser and callback behavior, interactivity controls, client-registration recovery, callback parsing, metadata handling, callback-port reservation, and failure cleanup.

The branch is rebased onto current `main` and preserves all newer upstream coverage for per-flow redirect-handler closures, configured redirect URIs, cached callback ports, and TOCTOU-safe socket reservation. The older tests were reconciled with those current contracts instead of restoring deprecated module-global redirect behavior.

## Related Issue

Fixes #36567

## Type of Change

- [x] Tests

## Changes Made

- Preserves current `main` tests for redirect URI proxying and per-flow callback-port isolation.
- Adds token, client-info, metadata, JSON persistence, snapshot/restore, browser detection, callback, paste-reader, and registration-recovery edge cases.
- Covers callback socket bind failure, reservation eviction cleanup, invalid cached redirect URIs, and clear port-collision reporting.
- Keeps non-interactive fail-fast tests separate from downstream interactive callback tests.
- Removes duplicate test-class shadowing so every intended interactivity regression is actually collected.

## How to Test

- [x] `scripts/check.sh --project hermes-agent --worktree worktrees/hermes-agent/36567`
- [x] 197 tests passed in `tests/tools/test_mcp_oauth.py`.
- [x] `tools/mcp_oauth.py`: 99% coverage, 440 statements, 4 missing optional import-fallback lines (`74-75`, `79-80`).
- [x] `ruff check .` passed.
- [x] `uv lock --check` passed.
- [x] Targeted `ty` reports the same six diagnostics as current `main`, with zero new diagnostics from this PR.
- [x] No production file, dependency, lockfile, or configuration changed.

## Checklist

- [x] The conflict was resolved against current `main` without dropping upstream tests.
- [x] All intended test classes and methods are collected exactly once.
- [x] Tests exercise real persistence, socket, callback, parser, and recovery behavior.
- [x] Coverage and test counts were remeasured on the exact final commit.
- [x] Maintainer edits remain enabled.

### CODE REVIEW

**Review by teknium1 — COMMENTED** [review]
Thanks for the extensive MCP OAuth coverage work. The test targets moved after this PR was updated.

### Problems
- Current main intentionally rejects non-interactive callback flows before creating `HTTPServer` or inspecting callback results (`tools/mcp_oauth.py:643-647`, added by `755194ffe`). The PR sets `_is_interactive` to `False` at `tests/tools/test_mcp_oauth.py:1451`, `:1476`, and `:1941` while expecting downstream callback behavior, so those cases no longer exercise their stated branches.
- The browser-path tests at `tests/tools/test_mcp_oauth.py:1369-1411` also need to establish an interactive context: `_redirect_handler()` now fails before browser handling when non-interactive (`tools/mcp_oauth.py:557-560`).

### Suggested changes
- Preserve a non-interactive regression asserting the callback guard fires before binding, and set `_is_interactive` to `True` in tests intended to cover callback/result/browser paths.
- Recalculate coverage after reconciling with the current source and its post-PR regression tests.

Automated hermes-sweeper review.

File: tests/tools/test_mcp_oauth.py [hunk 1]
```diff
@@ -858,3 +932,1028 @@ def test_poison_noop_when_no_client_file(self, tmp_path, monkeypatch):
...
+                asyncio.run(_wait_for_callback())
+        finally:
+            s.close()
+
+    def test_raises_runtime_error_on_auth_error(self, monkeypatch):
+        """result['error'] is set (non-skip) → RuntimeError."""
+        import tools.mcp_oauth as mod
+
+        mod._oauth_port = _find_free_port()
+        monkeypatch.setattr(mod, "_is_interactive", lambda: False)
```
Comment (teknium1) [inline tests/tools/test_mcp_oauth.py]: Current main now rejects non-interactive callback flows before `_make_callback_handler` or `HTTPServer` (`tools/mcp_oauth.py:643-647`). Set this to `True` if the intent is to cover the later `RuntimeError` result path; keep a separate false-context test for the fail-fast invariant.

File: tests/tools/test_mcp_oauth.py [hunk 2]
```diff
@@ -858,3 +932,1028 @@ def test_poison_noop_when_no_client_file(self, tmp_path, monkeypatch):
...
+
+
+class TestWaitForCallbackSuccessReturn:
+    """_wait_for_callback returns (auth_code, state) when result has auth_code."""
+
+    def test_returns_auth_code_and_state(self, monkeypatch):
+        import tools.mcp_oauth as mod
+
+        mod._oauth_port = _find_free_port()
+        monkeypatch.setattr(mod, "_is_interactive", lambda: False)
```
Comment (teknium1) [inline tests/tools/test_mcp_oauth.py]: With `_is_interactive=False`, current main raises at the callback boundary before the mocked success result can be returned (`tools/mcp_oauth.py:643-647`). Use an interactive context for this success-path test.

**Comment by Christopher-Schulze** [comment]
Reconciled the suite with the current non-interactive guard. Only tests targeting downstream browser, port-bind, callback-error, timeout, and success behavior now establish an interactive context; the current-main fail-fast tests remain intact and prove non-interactive flows do not open a browser or bind a listener. I also covered the newer snapshot/restore and forced-interactivity paths. The full file passes 165 tests at 99% coverage of the current 377-statement module, and all blocking contributor gates pass.

File: tests/tools/test_mcp_oauth.py (hunk 1 above)
Comment (Christopher-Schulze) [reply tests/tools/test_mcp_oauth.py]: Addressed: the later RuntimeError path now uses an interactive context; the separate non-interactive fail-fast test remains.

File: tests/tools/test_mcp_oauth.py (hunk 2 above)
Comment (Christopher-Schulze) [reply tests/tools/test_mcp_oauth.py]: Addressed: the callback success-path test now uses an interactive context.

**Comment by Christopher-Schulze** [comment]
Rebased onto current `main` and resolved the MCP OAuth test conflict without dropping the newer redirect-URI or callback-port reservation coverage. The final branch now collects 197 tests, passes the full focused file, and measures 99% coverage (440 statements, only the four optional import-fallback lines missing). The contributor harness blocking gates are green, and targeted `ty` shows zero diagnostics beyond the same six already present on `main`.
